src/agent/src/image_rpc.rs

```diff
@@ -33,11 +58,227 @@ pub struct ImageService {
+                info!(sl!(), "Saving layer file {}...", layer_name);
+                out_file.write_all(&layer.data)?;
+            } else {
+                error!(
+                    sl!(),
+                    "layer digest algo not supported:: {:?}", layer_digest
+                );
+            }
+        }
+
+        Ok(())
+    }
+
+    #[tokio::main]
+    async fn download_image(
+        image: &str,
+        auth: &RegistryAuth,
+    ) -> anyhow::Result<(OciManifest, String, ImageData)> {
+        let reference = Reference::try_from(image)?;
+        let mut client = Client::default();
+        let (image_manifest, _image_digest, image_config) =
+            client.pull_manifest_and_config(&reference, auth).await?;
+
+        let mut image_data = ImageData {
+            layers: Vec::with_capacity(0),
+            digest: None,
+        };
+
+        // TODO: Get the value from config
+        let max_attempt = 2;
+        let attempt_interval = 1;
+        for i in 1..max_attempt {
+            match client
+                .pull(
+                    &reference,
+                    auth,
+                    vec![
+                        manifest::IMAGE_DOCKER_LAYER_GZIP_MEDIA_TYPE,
+                        manifest::IMAGE_LAYER_GZIP_MEDIA_TYPE,
+                        MEDIA_TYPE_LAYER_GZIP_ENC,
+                        MEDIA_TYPE_LAYER_ENC,
+                        MEDIA_TYPE_LAYER_NON_DISTRIBUTABLE_ENC,
+                        MEDIA_TYPE_LAYER_NON_DISTRIBUTABLE_GZIP_ENC,
+                    ],
+                )
+                .await
+            {
+                Ok(data) => {
+                    image_data = data;
+                    break;
+                }
+                Err(e) => {
+                    info!(
+                        sl!(),
+                        "Got error on pull call attempt #{}. Will retry in {}s: {:?}",
+                        attempt_interval,
+                        i,
+                        e
+                    );
+                    tokio::time::sleep(tokio::time::Duration::from_secs(attempt_interval)).await;
+                }
+            }
+        }
+
+        Ok((image_manifest, image_config, image_data))
+    }
+
+    fn pull_image_with_oci_distribution(
+        image: &str,
+        cid: &str,
+        source_creds: &Option<String>,
+    ) -> Result<()> {
+        let oci_path = Self::build_oci_path(cid);
+        fs::create_dir_all(&oci_path)?;
+
+        info!(
+            sl!(),
+            "Attempting to pull image with native implementation {}...", image
+        );
+
+        let mut auth = RegistryAuth::Anonymous;
+        if let Some(source_creds) = source_creds {
+            let auth_info: Vec<&str> = source_creds.split(':').collect();
+            if auth_info.len() == 2 {
+                auth = RegistryAuth::Basic(auth_info[0].to_string(), auth_info[1].to_string());
+            } else {
+                warn!(sl!(), "invalid auth_info format");
```

Comment:
Should this return `Err()` instead? If not, can you change the message to make the behaviour clearer? Something like:

```suggestion
                warn!(sl!(), "ignoring invalid authentication info ({:?})", source_creds);
```

Reply:
thanks, this warning will be more clear